tests/python/relax/test_training_optimizer.py

```diff
@@ -0,0 +1,508 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Unit tests for relax optimizer APIs."""
+import pytest
+import tvm
+import tvm.testing
+from tvm import relax
+from tvm.ir.base import assert_structural_equal
+from tvm.relax.training.optimizer import SGD, MomentumSGD, Adam
+from tvm.script.parser import relax as R
+
+
+def test_optimizer_wrong_param():
+    x1 = relax.Var("x1", R.Tensor((3, 3), "float32"))
+    x2 = relax.Var("x1", R.Tensor((3, 3), "bfloat16"))
+    x3 = relax.Var("x2", R.Tuple([R.Tensor((3, 3), "float32")]))
+    x4 = relax.Var("x3", R.Tensor((3, 3), "int64"))
+    x5 = relax.Tuple([x1])
+
+    # fine cases
+    SGD(x1, 0.01)
+    SGD([x1], 0.01)
+    assert SGD([x2], 0.01)._dtype == "bfloat16"
+
+    with pytest.raises(AssertionError, match="There exist duplicate Vars in the parameter list."):
+        SGD([x1, x1], 0.01)
+    with pytest.raises(AssertionError, match="There exist duplicate Vars in the parameter list."):
+        SGD(x1, 0.01).add_params([x1])
+    with pytest.raises(AssertionError, match="Not every parameter is Var."):
+        SGD(x5, 0.01)
+    with pytest.raises(AssertionError, match="Not every parameter is Tensor Var"):
+        SGD(x3, 0.01)
+    with pytest.raises(AssertionError, match="Pamameters must be of float dtype"):
+        SGD(x4, 0.01)
+    with pytest.raises(AssertionError, match="All parameters should have the same dtype"):
+        SGD([x1, x2], 0.01)
+
+
+def test_sgd_simple():
+    x = relax.Var("x", R.Tensor((3, 3), "float32"))
+    y = relax.Var("y", R.Tensor((3,), "float32"))
+    sgd = SGD([x, y], 0.01).get_function()
+
+    # fmt: off
+    @R.function
+    def sgd_expected(params: R.Tuple(R.Tensor((3, 3), dtype="float32"), R.Tensor((3,), dtype="float32")), gradients: R.Tuple(R.Tensor((3, 3), dtype="float32"), R.Tensor((3,), dtype="float32")), optim_states: R.Tuple(R.Tensor((), dtype="int64"))) -> R.Tuple(R.Tuple(R.Tensor((3, 3), dtype="float32"), R.Tensor((3,), dtype="float32")), R.Tuple(R.Tensor((), dtype="int64"))):
```

Comment:
I think we can enable formatting in this file as these are the basic tests for optimizer, and the formatted code is crystal clear.

To avoid lines being too long, you can safely remove all `dtype=` in `R.Tensor(..)`, as the parser is able to identity `R.Tensor((3, 3), "float32")`.

Reply:
Do you mean removing `dtype` in parameters, or bindings?

It seems that parameter dtypes needs to be specified to deduce dtypes of following bindings

Reply:
I mean if you like you can write `R.Tensor((3, 3), "float32")`, which is 6 characters shorter than `R.Tensor((3, 3), dtype="float32")`.